diffie-hellman: Add test cases for inputs near the type limits

LeBlue opened the pull request:
Fixes https://github.com/exercism/rust/issues/936

* Add tests with inputs near the u64 type limits as a solutions may
  panic 'multiplication overflow' while passing existing tests
* Fix example code to also pass these tests

coriolinus commented on exercises/diffie-hellman/tests/diffie-hellman.rs:
```diff
@@ -67,6 +67,69 @@ fn test_secret_key_correct_big_numbers() {
+#[test]
+#[ignore]
+fn test_public_key_correct_biggest_numbers() {
+    // two biggest 64bit primes
+    let p: u64 = 0xF_FFF_FFF_FFF_FFF_FC5;
+    let g: u64 = 0xF_FFF_FFF_FFF_FFF_FAC;
+
+    let private_key: u64 = 0xF_FFF_FFF_FFF_FFF_FC3;
+
+    let expected: u64 = 13_281_655_733_070_877_121;
```
You define the first two of these values in three places, and the other two in two. May as well extract them all into module-private consts.

LeBlue replied:
Done

coriolinus commented on exercises/diffie-hellman/tests/diffie-hellman.rs:
```diff
@@ -67,6 +67,69 @@ fn test_secret_key_correct_big_numbers() {
+#[ignore]
+fn test_public_key_correct_biggest_numbers() {
+    // two biggest 64bit primes
+    let p: u64 = 0xF_FFF_FFF_FFF_FFF_FC5;
```
Just a style thing: with hexadecimal values, it is unconventional to divide the digits into groups of three, as you do here. It is much more common to group the digits to some power of 2, because every two hex digits corresponds to exactly one byte.

```suggestion
    let p: u64 = 0xFFFF_FFFF_FFFF_FFC5;
```

The same goes, of course, for all the other hex literals.

LeBlue replied:
Done

LeBlue commented:
@coriolinus I implemented the suggested changes.

For option
> Retain the example.rs changes, drop the diffie-hellman.rs changes, so that the example is correct but there are no tests verifying the behavior for large inputs.

I split of the example fixes in a separate commit (Do I need to/should I create a new pull request for that?).

The third option
> Keep all these changes, but feature-gate the new tests, adding an appropriate section to the README briefly describing the nature of the feature-gated tests and how to run them. Feature gates neatly get us out of the problem of correctness issues which are out of scope of the main learning objective for an exercise.

is now implemented in this pull request. I am not really satisfied with the section in the README, feel free to suggest some changes.

coriolinus commented on exercises/diffie-hellman/.meta/hints.md:
```diff
@@ -1,2 +1,17 @@
+
+## For bonus points
+
+Did you get the tests passing and the code clean? If you want to, there
+are some additional things you could try.
+
+- Currently your modular exponentiation function will probably not work for
+  the full range of inputs. There are additional tests you can run.
```
```suggestion
Many implementations do not work for
the entire domain of inputs. There are additional optional tests 
defined which help ensure that all valid inputs produce valid results.
```

LeBlue replied:
Included

coriolinus commented on exercises/diffie-hellman/Cargo-example.toml:
```diff
@@ -5,3 +5,6 @@ version = "0.1.0"
 
 [dependencies]
 rand = "0.3.18"
+
+[features]
+big-primes = []
```
Please ensure that all text files end with a newline. I see that the original `Cargo.toml` didn't, which was an error of review on our part.

LeBlue replied:
Also noticed that and decided to stick with the original

coriolinus commented on exercises/diffie-hellman/tests/diffie-hellman.rs:
```diff
@@ -67,6 +67,63 @@ fn test_secret_key_correct_big_numbers() {
 }
 
+// two biggest 64bit primes
+#[allow(dead_code)]
```
```suggestion
#[cfg(feature = "big-primes")]
```

These consts are only dead code when the appropriate tests are ignored during compilation, so may as well apply the same feature gate to the consts instead of compiling them unconditionally.

LeBlue replied:
I wondered if there was a better solution

LeBlue commented:
@coriolinus All suggestions should be included/all fixed now.

patricksjackson commented:
I'm not a maintainer here, but am on the C/C++ tracks.

I agree that keeping these changes as bonus tests is the best choice here, LGTM :+1:
